src/utils/viewer/viewer.cpp

```diff
@@ -552,11 +588,90 @@ void Viewer::viewportEvent(ViewportEvent& event) {
+  selectionDrawableId_.setStorage(Mn::GL::RenderbufferFormat::R32UI,
+                                  event.framebufferSize());
+  selectionFramebuffer_
+      .attachRenderbuffer(Mn::GL::Framebuffer::BufferAttachment::Depth,
+                          selectionDepth_)
+      .attachRenderbuffer(Mn::GL::Framebuffer::ColorAttachment{1},
+                          selectionDrawableId_)
+      .setViewport({{}, event.framebufferSize()});
 }
 
 void Viewer::mousePressEvent(MouseEvent& event) {
+  if (event.button() == MouseEvent::Button::Right && objectSelectionOn_) {
+    selectionFramebuffer_.bind();
+    selectionFramebuffer_
+        .mapForDraw({{Mn::Shaders::Generic3D::ColorOutput,
+                      Mn::GL::Framebuffer::DrawAttachment::None},
+                     {Mn::Shaders::Generic3D::ObjectIdOutput,
+                      Mn::GL::Framebuffer::ColorAttachment{1}}})
+        .clearDepth(1.0f)
+        .clearColor(1, Mn::Vector4ui{0xffff});
+    CORRADE_INTERNAL_ASSERT(
+        selectionFramebuffer_.checkStatus(Mn::GL::FramebufferTarget::Draw) ==
+        Mn::GL::Framebuffer::Status::Complete);
+
+    int DEFAULT_SCENE = 0;
+    auto& sceneGraph = sceneManager_.getSceneGraph(sceneID_[DEFAULT_SCENE]);
+
+    // if there is a selected object, remove it first
+    if (selectedDrawableId_ >= 0) {
+      for (std::pair<std::string, esp::gfx::DrawableGroup>&& it :
+           sceneGraph.getDrawableGroups()) {
+        esp::gfx::Drawable* d = it.second.getDrawable(selectedDrawableId_);
+        if (d != nullptr) {
+          d->setSelected(false);
+          break;  // the selected object is found and processed
+        }
+      }
+      selectedDrawableId_ = -1;
+    }
+    for (std::pair<std::string, esp::gfx::DrawableGroup>&& it :
+         sceneGraph.getDrawableGroups()) {
+      renderCamera_->draw(it.second, frustumCullingEnabled_);
+    }
+
+    // Read the ID back
+    selectionFramebuffer_.mapForRead(Mn::GL::Framebuffer::ColorAttachment{1});
+    CORRADE_INTERNAL_ASSERT(
+        selectionFramebuffer_.checkStatus(Mn::GL::FramebufferTarget::Read) ==
+        Mn::GL::Framebuffer::Status::Complete);
+
+    /* First scale the position from being relative to window size to being
+       relative to framebuffer size as those two can be different on HiDPI
+       systems */
+    const Mn::Vector2i position = event.position() *
+                                  Mn::Vector2{framebufferSize()} /
+                                  Mn::Vector2{windowSize()};
+    const Mn::Vector2i fbPosition{
+        position.x(),
+        selectionFramebuffer_.viewport().sizeY() - position.y() - 1};
+    const Mn::Range2Di area =
+        Mn::Range2Di::fromSize(fbPosition, Mn::Vector2i{1});
+
+    const Mn::UnsignedInt selectedObject =
+        selectionFramebuffer_.read(area, {Mn::PixelFormat::R32UI})
+            .pixels<Mn::UnsignedInt>()[0][0];
+
+    for (std::pair<std::string, esp::gfx::DrawableGroup>&& it :
```

Comment:
```suggestion
    for (std::pair<const std::string, esp::gfx::DrawableGroup>&& it :
```

or this is at least what latest Clang is always pointing out, that you'd get a copy otherwise (oh, wait, that means `DrawableGroup` is *copyable*? :scream:)

Same in all cases above.

Reply:
@mosra: Oh, I see. I introduced a bug here. If I use `auto&&` directly, it should be prevented.

The container for a STL `map` or `unordered_map` is actually `std::pair<const K, V>` NOT `std::pair<K, V>`.

Am I right? Wow, good catch.